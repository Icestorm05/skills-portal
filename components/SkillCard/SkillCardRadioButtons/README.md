# Skill Card Radio Buttons

## Functionality

The skill card radio buttons component displays two sets of radio button groups for skill and currency. The value of the initial radio button is determined by the values passed in.

## Usage

### Props

`<skill-card>` supports the following custom component attributes:

| attribute | type | description
| --- | --- | ---
| `currency` | Number OR Null *optional* | The skill currency ID. Can use the `sync` operator, for a two-way data binding. Defaults to `null`.
| `disabled` | Boolean *optional* | Whether the radio buttons are disabled or not. Defaults to `false`.
| `level` | Number OR Null *optional* | The skill level ID. Can use the `sync` operator, for a two-way data binding. Defaults to `null`.

### Injections

`<skill-card-radio-buttons>` requires a parent component to provide the following:

| injection | type | description
| --- | --- | ---
| validator | Validator | The VeeValidate $validator instance. The parent component that will check the validation of this component should provide this.

### Example

```html
<skill-card-radio-buttons :currency.sync="currency"
                          :level.sync="level"/>
```

```ts
import Component from 'nuxt-class-component';
import Vue from 'vue';
import { Provide } from 'vue-property-decorator';

import SkillCardRadioButtons from '~/components/SkillCard/SkillCardRadioButtons/SkillCardRadioButtons.vue';

@Component({
    components: {
        SkillCardRadioButtons,
    },
    // Initialize a fresh validator instance.
    $_veeValidate: {
        validator: 'new',
    },
})
export default class GenericComponent extends Vue {
    // Provide the fresh validator instance for <skill-card-status> to inject.
    @Provide('validator') public $validator = this.$validator;

    private currency = 4;
    private level = 2;
}
```