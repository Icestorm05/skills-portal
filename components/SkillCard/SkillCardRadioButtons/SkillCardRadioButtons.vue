<template lang="pug">
    v-layout(full-width).SkillCardRadioButtons
        v-flex(xs6)
            v-radio-group(:input-value="level"
                          @change="$emit('update:level', $event)"
                            v-validate:level="'required'"
                            label="Level"
                            name="level"
                            :error-messages="errors.collect('level')"
                            :disabled="disabled").SkillCardRadioGroup
                v-radio(:value="4"
                        label="Expert"
                        color="success"
                        light).SkillCardRadioButton
                v-radio(:value="3"
                        label="Practitioner"
                        color="warning"
                        light).SkillCardRadioButton
                v-radio(:value="2"
                        label="Awareness"
                        color="error"
                        light).SkillCardRadioButton
        v-flex(xs6)
            v-radio-group(:input-value="currency"
                            @change="$emit('update:currency', $event)"
                            v-validate:currency="'required'"
                            label="Currency"
                            name="currency"
                            :error-messages="errors.collect('currency')"
                            :disabled="disabled").SkillCardRadioGroup
                v-radio(:value="2"
                        label="Current"
                        color="success"
                        light).SkillCardRadioButton
                v-radio(:value="3"
                        label="Fading"
                        color="warning"
                        light).SkillCardRadioButton
                v-radio(:value="4"
                        label="Historic"
                        color="error"
                        light).SkillCardRadioButton
</template>

<script lang="ts">
import Component from "nuxt-class-component";
import { Validator } from "vee-validate";
import Vue from "vue";
import { Inject, Prop } from "vue-property-decorator";

import { VFlex, VLayout } from "vuetify/es5/components/VGrid";
import VRadioGroup, { VRadio } from "vuetify/es5/components/VRadioGroup";

@Component({
  components: {
    VFlex,
    VLayout,
    VRadioGroup,
    VRadio
  }
})
export default class SkillCardRadioButtons extends Vue {
  @Inject("validator") public $validator: Validator;
  @Prop({ type: Boolean, default: false })
  private disabled: boolean;
  @Prop({
    default: null,
    validator: val => val === null || typeof val === "number"
  })
  private currency: number | null;
  @Prop({
    default: null,
    validator: val => val === null || typeof val === "number"
  })
  private level: number | null;
}
</script>
