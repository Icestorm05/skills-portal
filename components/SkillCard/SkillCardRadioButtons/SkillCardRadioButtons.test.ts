import SkillCardRadioButtons from "./SkillCardRadioButtons.vue";
import VeeValidate from "vee-validate";
import { createLocalVue, mount } from "@vue/test-utils";

declare const expect: jest.Expect;

const validator = new VeeValidate.Validator({}, {});

const localVue = createLocalVue();
localVue.use(VeeValidate);

describe("Skill Card Radio Buttons", () => {
  test("is a Vue instance", () => {
    const wrapper = mount(SkillCardRadioButtons, {
      localVue,
      provide: { validator }
    });
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test("snapshot has not changed", async () => {
    const wrapper = mount(SkillCardRadioButtons, {
      localVue,
      propsData: { currency: 2, level: 4 },
      provide: { validator }
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("currency and level prop are null by default", () => {
    const wrapper = mount(SkillCardRadioButtons, {
      localVue,
      provide: { validator }
    });
    expect(wrapper.vm.$props.currency).toEqual(null);
    expect(wrapper.vm.$props.level).toEqual(null);
  });

  test("currency and level prop can be set to null", () => {
    const wrapper = mount(SkillCardRadioButtons, {
      localVue,
      propsData: { currency: null, level: null },
      provide: { validator }
    });
    expect(wrapper.vm.$props.currency).toEqual(null);
    expect(wrapper.vm.$props.level).toEqual(null);
  });

  test("setting currency and level prop overrides default", () => {
    const wrapper = mount(SkillCardRadioButtons, {
      localVue,
      propsData: { currency: 2, level: 4 },
      provide: { validator }
    });
    expect(wrapper.vm.$props.currency).toEqual(2);
    expect(wrapper.vm.$props.level).toEqual(4);
  });

  test("radio buttons are unselected by default", async () => {
    const wrapper = mount(SkillCardRadioButtons, {
      localVue,
      provide: { validator }
    });
    await wrapper.vm.$nextTick();
    expect(
      wrapper.findAll(".SkillCardRadioButton.input-group--active").exists()
    ).toBeFalsy();
  });

  test("currency radio button is selected by default with currency prop", async () => {
    const wrapper = mount(SkillCardRadioButtons, {
      localVue,
      propsData: { currency: 2 },
      provide: { validator }
    });
    await wrapper.vm.$nextTick();
    const selectedRadioButtons = wrapper.findAll(
      ".SkillCardRadioButton.input-group--active"
    );
    const selectedCurrency = wrapper.find(
      ".SkillCardRadioButton.input-group--active input"
    ).element as HTMLInputElement;
    expect(selectedRadioButtons.exists()).toBeTruthy();
    expect(selectedRadioButtons.length).toEqual(1);
    expect(selectedCurrency.value).toEqual("2");
  });

  test("level radio button is selected by default with level prop", async () => {
    const wrapper = mount(SkillCardRadioButtons, {
      localVue,
      propsData: { level: 4 },
      provide: { validator }
    });
    await wrapper.vm.$nextTick();
    const selectedRadioButtons = wrapper.findAll(
      ".SkillCardRadioButton.input-group--active"
    );
    const selectedLevel = wrapper.find(
      ".SkillCardRadioButton.input-group--active input"
    ).element as HTMLInputElement;
    expect(selectedRadioButtons.exists()).toBeTruthy();
    expect(selectedRadioButtons.length).toEqual(1);
    expect(selectedLevel.value).toEqual("4");
  });

  test("changing currency prop changes the radio button selected", async () => {
    const wrapper = mount(SkillCardRadioButtons, {
      localVue,
      propsData: { currency: 2 },
      provide: { validator }
    });
    wrapper.setProps({ currency: 3 });
    await wrapper.vm.$nextTick();
    const selectedLevel = wrapper.find(
      ".SkillCardRadioButton.input-group--active input"
    ).element as HTMLInputElement;
    expect(selectedLevel.value).toEqual("3");
  });

  test("changing level prop changes the radio button selected", async () => {
    const wrapper = mount(SkillCardRadioButtons, {
      localVue,
      propsData: { level: 4 },
      provide: { validator }
    });
    wrapper.setProps({ level: 3 });
    await wrapper.vm.$nextTick();
    const selectedLevel = wrapper.find(
      ".SkillCardRadioButton.input-group--active input"
    ).element as HTMLInputElement;
    expect(selectedLevel.value).toEqual("3");
  });
});
