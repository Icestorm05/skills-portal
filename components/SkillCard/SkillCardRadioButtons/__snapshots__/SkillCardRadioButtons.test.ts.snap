// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skill Card Radio Buttons snapshot has not changed 1`] = `
<div class="layout SkillCardRadioButtons full-width">
    <div class="flex xs6">
        <div role="radiogroup" class="SkillCardRadioGroup input-group input-group--dirty radio-group radio-group--column" data-vv-id="3" aria-required="true" aria-invalid="false">
            <label>Level</label>
            <div class="input-group__input">
                <div role="radio" aria-checked="true" aria-label="Expert" class="SkillCardRadioButton input-group input-group--active input-group--selection-controls radio theme--light success--text" tabindex="0">
                    <label>Expert</label>
                    <div class="input-group__input">
                        <i aria-hidden="true" class="icon icon--selection-control material-icons fade-transition-leave fade-transition-leave-active">radio_button_unchecked</i>
                        <i aria-hidden="true" class="icon icon--selection-control material-icons icon--radio fade-transition-enter fade-transition-enter-active" style="-webkit-transform-origin: top center 0;">radio_button_checked</i>
                        <div class="input-group--selection-controls__ripple"></div>
                        <input name="level" type="radio" value="4" style="display: none;">
                    </div>
                </div>
                <div role="radio" aria-checked="false" aria-label="Practitioner" class="SkillCardRadioButton input-group input-group--selection-controls radio theme--light warning--text" tabindex="0">
                    <label>Practitioner</label>
                    <div class="input-group__input">
                        <i aria-hidden="true" class="icon icon--selection-control material-icons">radio_button_unchecked</i>
                        <div class="input-group--selection-controls__ripple"></div>
                        <input name="level" type="radio" value="3" style="display: none;">
                    </div>
                </div>
                <div role="radio" aria-checked="false" aria-label="Awareness" class="SkillCardRadioButton input-group input-group--selection-controls radio theme--light error--text" tabindex="0">
                    <label>Awareness</label>
                    <div class="input-group__input">
                        <i aria-hidden="true" class="icon icon--selection-control material-icons">radio_button_unchecked</i>
                        <div class="input-group--selection-controls__ripple"></div>
                        <input name="level" type="radio" value="2" style="display: none;">
                    </div>
                </div>
            </div>
            <div class="input-group__details">
                <!---->
            </div>
        </div>
    </div>
    <div class="flex xs6">
        <div role="radiogroup" class="SkillCardRadioGroup input-group input-group--dirty radio-group radio-group--column" data-vv-id="4" aria-required="true" aria-invalid="false">
            <label>Currency</label>
            <div class="input-group__input">
                <div role="radio" aria-checked="true" aria-label="Current" class="SkillCardRadioButton input-group input-group--active input-group--selection-controls radio theme--light success--text" tabindex="0">
                    <label>Current</label>
                    <div class="input-group__input">
                        <i aria-hidden="true" class="icon icon--selection-control material-icons fade-transition-leave fade-transition-leave-active">radio_button_unchecked</i>
                        <i aria-hidden="true" class="icon icon--selection-control material-icons icon--radio fade-transition-enter fade-transition-enter-active" style="-webkit-transform-origin: top center 0;">radio_button_checked</i>
                        <div class="input-group--selection-controls__ripple"></div>
                        <input name="currency" type="radio" value="2" style="display: none;">
                    </div>
                </div>
                <div role="radio" aria-checked="false" aria-label="Fading" class="SkillCardRadioButton input-group input-group--selection-controls radio theme--light warning--text" tabindex="0">
                    <label>Fading</label>
                    <div class="input-group__input">
                        <i aria-hidden="true" class="icon icon--selection-control material-icons">radio_button_unchecked</i>
                        <div class="input-group--selection-controls__ripple"></div>
                        <input name="currency" type="radio" value="3" style="display: none;">
                    </div>
                </div>
                <div role="radio" aria-checked="false" aria-label="Historic" class="SkillCardRadioButton input-group input-group--selection-controls radio theme--light error--text" tabindex="0">
                    <label>Historic</label>
                    <div class="input-group__input">
                        <i aria-hidden="true" class="icon icon--selection-control material-icons">radio_button_unchecked</i>
                        <div class="input-group--selection-controls__ripple"></div>
                        <input name="currency" type="radio" value="4" style="display: none;">
                    </div>
                </div>
            </div>
            <div class="input-group__details">
                <!---->
            </div>
        </div>
    </div>
</div>
`;
